File: solutions/142/main.cpp
#include <gtest/gtest.h> 

using namespace std; 

// Definition for singly-linked list.
struct ListNode {
    int val; 
    ListNode* next; 
    ListNode(int x) : val(x), next(nullptr) {}
}; 

// time complexity O(m+n)
class Solution {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB, ListNode *pseudoEnd) {
        ListNode* curA = headA; 
        ListNode* curB = headB; 
        while (curA != curB) {
            curA = curA == pseudoEnd ? headB : curA->next; 
            curB = curB == pseudoEnd ? headA : curB->next; 
        }

        return curA; 
    }

    ListNode *detectCycle(ListNode *head) {
        ListNode* fast = head; 
        ListNode* slow = head; 
        while (fast != nullptr && fast->next != nullptr && fast->next != slow) {
            fast = fast->next->next; 
            slow = slow->next; 
        }
        if (fast == nullptr || fast->next == nullptr)
            return nullptr; 
        else {
            ListNode* pseudoEnd = fast;
            return getIntersectionNode(head, slow, pseudoEnd); 
        }
    }
};

TEST(Test142, SimpleTest) {
    ListNode* headA = new ListNode(3);
    ListNode* n1A = new ListNode(2);
    ListNode* n2A = new ListNode(0);
    ListNode* n3A = new ListNode(-4);

    headA->next = n1A; 
    n1A->next = n2A; 
    n2A->next = n3A; 
    n3A->next = n1A; 

    Solution s; 
    ListNode* inter = s.detectCycle(headA);

    EXPECT_EQ(inter, n1A); 

    delete headA;
    delete n1A;  
    delete n2A;
    delete n3A;
}

int main() {
    testing::InitGoogleTest(); 
    return RUN_ALL_TESTS();
}

File: solutions/142/main.py
from typing import * 

# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def detectCycle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        slow, fast = head, head 
        while fast is not None and fast.next is not None and fast.next != slow: 
            slow = slow.next 
            fast = fast.next.next 
        # no cycle 
        if fast is None or fast.next is None: 
            return None 
        
        # find intersection, ref 160 
        # one linked list starts at head, one starts at slow 
        cur1, cur2 = head, slow 
        end = fast 
        while cur1 != cur2: 
            if cur1 == end: 
                cur1 = slow 
            else: 
                cur1 = cur1.next 
            if cur2 == end: 
                cur2 = head 
            else: 
                cur2 = cur2.next 
            
        return cur1 